mixnet/client/src/sender.rs

```diff
@@ -133,21 +133,60 @@ where
             return None;
         }
 
+        // corner case: 0 average delay
+        if self.avg_delay == 0 {
+            self.remaining_delays -= 1;
+            return Some(0);
+        }
+
         if self.remaining_delays == 1 {
             self.remaining_delays -= 1;
             return Some(self.remaining_time);
         }
 
+        // if we have no remaining time, then just return 0
+        if self.remaining_time == 0 {
+            self.remaining_delays -= 1;
+            return Some(0);
+        }
+
         // Calculate bounds to avoid extreme values
         let lower_bound = (self.avg_delay as f64 * 0.5) as u64;
         let upper_bound = (self.avg_delay as f64 * 1.5)
-            .min((self.remaining_time - self.remaining_delays + 1) as f64)
+            .min((self.remaining_time as f64 - self.remaining_delays as f64) + 1.0)
```

Comment:
not sure I understand this formula

Reply:
The line

```rust
.min((self.remaining_time - self.remaining_delays + 1) as f64)
```

is included to ensure that `upper_bound` never exceeds the maximum possible delay that can be returned without causing future delays to be zero or negative.

`self.remaining_time` is the total time still available for all remaining delays, and `self.remaining_delays` is the number of remaining delays, then the expression `(self.remaining_time - self.remaining_delays + 1)` calculates the maximum time we can allocate for the current delay without making future delays invalid (i.e., zero or negative).

For example, consider a case where we have `remaining_time = 100` and `remaining_delays = 10`. If we use up 95 time units for the current delay, then we would be left with only 5-time units to distribute among the remaining 9 delays, which is not possible without using zero or negative delays.

Here's how the logic works:

- `self.remaining_time` is the remaining time that needs to be allocated across all remaining delays.
- `self.remaining_delays` is the number of remaining delays.

The maximum you could use for the current delay to be still able to allocate at least 1 unit of time for each of the remaining delays is `(self.remaining_time - self.remaining_delays + 1)`.

The `min` function is used to choose the smaller of the two computed upper bounds (`1.5 * avg_delay` and `(self.remaining_time - self.remaining_delays + 1)`) to make sure that the program doesn't allocate too much time for the current delay.

Reply:
Got it, better to put this in a comment as well (it's enought to point out that we don't want future delays to be zero)